src/core/gui/OpacityPreviewToolbox.cpp

```diff
@@ -0,0 +1,531 @@
+        FloatingToolbox* oFloatingToolbox = self->theMainWindow->getFloatingToolbox();
+        if (!isPointerOverWidget(static_cast<int>(event->x_root), static_cast<int>(event->y_root),
+                                 oFloatingToolbox->floatingToolbox)) {
+            oFloatingToolbox->hide();
+        }
+
+        self->hideToolbox();
+    }
+    return false;
+}
+
+bool OpacityPreviewToolbox::isEnabled() {
+    bool result;
+
+    switch (toolHandler->getToolType()) {
+        case TOOL_PEN:
+            result = toolHandler->getPenFillEnabled() ? true : false;
+            break;
+        case TOOL_HIGHLIGHTER:
+            result = toolHandler->getHighlighterFillEnabled() ? true : false;
+            break;
+        case TOOL_SELECT_PDF_TEXT_RECT:
+        case TOOL_SELECT_PDF_TEXT_LINEAR:
+            result = true;
+            break;
+        default:
+            result = false;
+            break;
+    }
+    return result;
+}
+
+void OpacityPreviewToolbox::updateColor() {
+    this->color = theMainWindow->getControl()->getToolHandler()->getColor();
+
+    ToolType tooltype = toolHandler->getToolType();
+
+    switch (tooltype) {
+        case TOOL_PEN:
+            this->color.alpha = static_cast<uint8_t>(toolHandler->getPenFill());
+            break;
+        case TOOL_HIGHLIGHTER:
+            this->color.alpha = static_cast<uint8_t>(toolHandler->getHighlighterFill());
+            break;
+        default:
+            break;
+    }
+}
+
+void OpacityPreviewToolbox::update() {
+    if (this->toolHandler == nullptr) {
+        this->toolHandler = theMainWindow->getControl()->getToolHandler();
+    }
+    bool enabled = this->isEnabled();
+
+    if (enabled) {
+        this->resetEventBoxes();
+    } else {
+        for (EventBox eventBox: this->eventBoxes) {
+            gtk_widget_hide(eventBox.widget);
+        }
+        this->hideToolbox();
+    }
+}
+
+void OpacityPreviewToolbox::hideEventBoxesInFloatingToolBox() {
+    for (EventBox eventBox: this->eventBoxes) {
+        if (eventBox.inFloatingToolbox == true) {
+            gtk_widget_hide(eventBox.widget);
+        }
+    }
+}
+
+void OpacityPreviewToolbox::showEventBoxesInFloatingToolBox() {
+    for (EventBox eventBox: this->eventBoxes) {
+        if (eventBox.inFloatingToolbox == true) {
+            gtk_widget_hide(eventBox.widget);
+            gtk_widget_show_all(eventBox.widget);
+        }
+    }
+}
+
+static bool inline isAncestor(GtkWidget* ancestor, GtkWidget* child) {
+    GtkWidget* it = child;
+
+    while (it != nullptr) {
+        if (it == ancestor) {
+            return true;
+        }
+        it = gtk_widget_get_parent(it);
+    }
+    return false;
+}
+
+// Create and position an eventbox over each ColorToolItem matching
+// the active color and connect enter/leave handlers.
+void OpacityPreviewToolbox::resetEventBoxes() {
+    for (EventBox eventBox: this->eventBoxes) {
+        gtk_widget_destroy(eventBox.widget);
```

Comment:
I'm not sure about what I am supposed to do.
I've replaced :
```c++
struct EventBox {
...
GtkWidget* widget;
...
};
```
with
```c++
struct EventBox {
...
xoj::util::WidgetSPtr widget;
...
};
```
Then I have replaced all `eventbox.widget` by `eventbox.widget.get()` and `eventBox.widget = gtk_event_box_new();` by `eventBox.widget.reset(gtk_event_box_new(), xoj::util::refsink));`

In the lines above that you've reviewed (243-246), I've replaced `gtk_widget_destroy(eventBox.widget);` by `eventBox.widget.reset();` but I feel like I am also supposed to remove it from the overlay using `gtk_overlay_remove_overlay(overlay.get(), eventBox.widget.get()) (GTK4)` / `gtk_container_remove(overlay.get(), eventBox.widget.get()) (GTK3)` ? In that case, would it not suffice to itself to remove the eventbox from the overlay without using `xoj::util::WidgetSPtr` and its reset() method?

What is the advantage of using xoj::util::WidgetSPtr?

Reply:
Oh thank you, I understand now better the use of WidgetSPtr.

Concerning your question, I am not sure how this would be implemented. On which event would this widget be moved around?

Currently, when a ColorItem is selected, an eventbox is created and overlaid on it and other ColorItems that match the same color.

Do you think about a unique eventbox that would be overlaid when the mouse enters the ColorItem?

I am not sure if there was something wrong in my implementation but when I had tried something similar, I could not get the ColorItem trigger any enter/leave event when it was in the pressed state (unlike when in released state). I was not sure if this was a bug, feature of GTK or if I had not use the proper code.

Reply:
I've made some new attempts to connect handlers to the enter/leave signals on ColorToolItems, and it appears that there's nothing preventing it from functioning as expected. I'm not sure what was wrong with my previous implementation. It's possible that I was trying to connect to the GtkToolItem instead of the GtkButton, which could explain why it failed in my previous attempts. In this case, it seems feasible to use a unique box event, and even better, to get rid of it, which would make it easier for a GTK4 port

Reply:
What would be the best approach?
- Removing old and Connecting new enter/leave handlers on ColorToolItems that match only the newly selected color.
- Connecting enter/leave handlers on all ColorToolItems when loading toolbars. When the enter/leave event is triggered, handlers check if the concerned widget match the active color to decide if it should show/hide the opacity toolbox

Reply:
"Would that be fine to add a reference to the OpacityToolbox in each ColorToolItem for handling enter/leave events on these buttons, or could this lead to unintended coupling that might not align with the original design?"

Reply:
I don't use EventBox anymore